src/client.rs

```diff
@@ -280,3 +270,368 @@ where
+    pub async fn wait_for_finalized(
+        mut self,
+    ) -> Result<TransactionInBlock<'client, T>, Error> {
+        while let Some(status) = self.next().await? {
+            match status {
+                // Finalized! Return.
+                TransactionProgressStatus::Finalized(s) => return Ok(s),
+                // Error scenarios; return the error.
+                TransactionProgressStatus::FinalityTimeout(_) => {
+                    return Err(TransactionError::FinalitySubscriptionTimeout.into())
+                }
+                TransactionProgressStatus::Invalid => {
+                    return Err(TransactionError::Invalid.into())
+                }
+                TransactionProgressStatus::Usurped(_) => {
+                    return Err(TransactionError::Usurped.into())
+                }
+                TransactionProgressStatus::Dropped => {
+                    return Err(TransactionError::Dropped.into())
+                }
+                // Ignore and wait for next status event:
+                _ => continue,
+            }
+        }
+        Err(RpcError::Custom("RPC subscription dropped".into()).into())
+    }
+
+    /// Wait for the transaction to be finalized, and for the transaction events to indicate
+    /// that the transaction was successful. Returns the events associated with the transaction,
+    /// as well as a couple of other details (block hash and extrinsic hash).
+    ///
+    /// **Note:** consumes self. If you'd like to perform multiple actions as progress is made,
+    /// use [`TransactionProgress::next()`] instead.
+    pub async fn wait_for_finalized_success(self) -> Result<TransactionEvents<T>, Error> {
+        let evs = self.wait_for_finalized().await?.wait_for_success().await?;
+        Ok(evs)
+    }
+}
+
+/// Possible transaction progess states returned from our [`TransactionProgress::next()`] call.
+#[derive(Debug)]
+pub enum TransactionProgressStatus<'client, T: Config> {
+    /// Transaction is part of the future queue.
+    Future,
+    /// Transaction is part of the ready queue.
+    Ready,
+    /// The transaction has been broadcast to the given peers.
+    Broadcast(Vec<String>),
+    /// Transaction has been included in block with given hash.
+    InBlock(TransactionInBlock<'client, T>),
+    /// The block this transaction was included in has been retracted.
+    Retracted(T::Hash),
+    /// Maximum number of finality watchers has been reached,
+    /// old watchers are being removed.
+    FinalityTimeout(T::Hash),
+    /// Transaction has been finalized by a finality-gadget, e.g GRANDPA
+    Finalized(TransactionInBlock<'client, T>),
+    /// Transaction has been replaced in the pool, by another transaction
+    /// that provides the same tags. (e.g. same (sender, nonce)).
+    Usurped(T::Hash),
+    /// Transaction has been dropped from the pool because of the limit.
+    Dropped,
+    /// Transaction is no longer valid in the current state.
```

Comment:
When can this happen? Worth mentioning in the docs?

Reply:
Hopefully 5f9ce69e52a3987261d4ae14e6c9e736ba561664 helps, but I've still no idea when specifically a transaction can be deemed invalid; presumably the signature doesnt line up or some other thing like that, and it's possible that even if this is the case, once a runtime upgrade happens it might no longer be invalid and might still get included in a block!